pkg/controllers/nodeclaim/disruption/underutilization.go

```diff
@@ -0,0 +1,98 @@
+*/
+
+package disruption
+
+import (
+	"context"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/samber/lo"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	v1 "sigs.k8s.io/karpenter/pkg/apis/v1"
+	"sigs.k8s.io/karpenter/pkg/controllers/state"
+	"sigs.k8s.io/karpenter/pkg/metrics"
+	nodeclaimutil "sigs.k8s.io/karpenter/pkg/utils/nodeclaim"
+)
+
+// Underutilization is a nodeclaim sub-controller that adds or removes status conditions on empty nodeclaims based on TTLSecondsAfterUnderutilized
+type Underutilization struct {
+	kubeClient client.Client
+	cluster    *state.Cluster
+}
+
+//nolint:gocyclo
+func (e *Underutilization) Reconcile(ctx context.Context, nodePool *v1.NodePool, nodeClaim *v1.NodeClaim) (reconcile.Result, error) {
+	hasUnderutilizedCondition := nodeClaim.StatusConditions().Get(v1.ConditionTypeUnderutilized) != nil
+
+	// 1. If ConsolidationPolicyWhenUnderutilized is not configured or ConsolidateAfter isn't configured, remove the underutilization status condition
+	if nodePool.Spec.Disruption.ConsolidationPolicy != v1.ConsolidationPolicyWhenUnderutilized ||
+		nodePool.Spec.Disruption.ConsolidateAfter == nil ||
+		nodePool.Spec.Disruption.ConsolidateAfter.Duration == nil {
+		if hasUnderutilizedCondition {
+			_ = nodeClaim.StatusConditions().Clear(v1.ConditionTypeUnderutilized)
+			log.FromContext(ctx).V(1).Info("removing underutilization status condition, underutilization is disabled")
+		}
+		return reconcile.Result{}, nil
+	}
+	// 2. If NodeClaim is not initialized, remove the underutilization status condition
+	if !nodeClaim.StatusConditions().Get(v1.ConditionTypeInitialized).IsTrue() {
+		if hasUnderutilizedCondition {
+			_ = nodeClaim.StatusConditions().Clear(v1.ConditionTypeUnderutilized)
+			log.FromContext(ctx).V(1).Info("removing underutilization status condition, isn't initialized")
+		}
+		return reconcile.Result{}, nil
+	}
+	// Get the node to check for pods scheduled to it
+	n, err := nodeclaimutil.NodeForNodeClaim(ctx, e.kubeClient, nodeClaim)
+	if err != nil {
+		// 3. If Node mapping doesn't exist, remove the underutilization status condition
+		if nodeclaimutil.IsDuplicateNodeError(err) || nodeclaimutil.IsNodeNotFoundError(err) {
+			_ = nodeClaim.StatusConditions().Clear(v1.ConditionTypeUnderutilized)
+			if hasUnderutilizedCondition {
+				log.FromContext(ctx).V(1).Info("removing underutilization status condition, doesn't have a node mapping")
+			}
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	// If node is not utilized, clear the condition, and requeue when we expect the node to be underutilized
+	underutilized, timeLeft, reason := e.cluster.IsNodeUnderutilized(n.Spec.ProviderID, lo.FromPtr(nodePool.Spec.Disruption.ConsolidateAfter.Duration))
+	if !underutilized {
+		_ = nodeClaim.StatusConditions().Clear(v1.ConditionTypeUnderutilized)
```

Comment:
There seems to be a little bit of inconsistency between us calling Clear() inside of our if conditionals and calling Clear() outside. Ideally, we should be consistent with this everywhere. Since this is an idempotent call, seems like it makes sense to me that you would just always clear the condition and then only log when there is a change